src/charm.py

```diff
@@ -205,6 +225,26 @@ def __init__(self, *args, **kargs) -> None:
         )
         self.framework.observe(self.on.update_status, self._on_update_status)
 
+    def setup_state(self) -> None:
+        """Set up the charm state."""
+        try:
+            self.state = State.from_charm(self)
+        except CharmConfigInvalidError as exc:
+            self.unit.status = ops.BlockedStatus(exc.msg)
+            raise ConfigurationError(exc.msg) from exc
+
+        if proxy_config := self.state.proxy_config:
+            if http_proxy := proxy_config.http:
+                self.proxies["http"] = str(http_proxy)
+            if https_proxy := proxy_config.https:
+                self.proxies["https"] = str(https_proxy)
+            # there's no need for no_proxy if there's no http_proxy or https_proxy
+            no_proxy = proxy_config.no_proxy
+            if (https_proxy or http_proxy) and no_proxy:
+                self.proxies["no_proxy"] = no_proxy
+            if proxy_config.use_aproxy:
+                self.proxies["aproxy_address"] = proxy_config.aproxy_address
+
```

Comment:
Just out of curiosity, may I ask why proxy is handled specially?

Reply:
The usage of `self.proxies` was removed during refactor, but I forgot to delete this section.